File: src/main/java/org/tindalos/principle/app/service/Application.java
package org.tindalos.principle.app.service;

import org.tindalos.principle.domain.checker.DesignQualityCheckResults;
import org.tindalos.principle.domain.checker.DesignQualityCheckService;
import org.tindalos.principle.domain.core.DesignQualityCheckConfiguration;
import org.tindalos.principle.domain.resultprocessing.reporter.DesignQualityCheckResultsReporter;
import org.tindalos.principle.domain.resultprocessing.reporter.Printer;
import org.tindalos.principle.domain.resultprocessing.thresholdchecker.ThresholdTrespassedException;

public class Application {
	
	private final DesignQualityCheckService designQualityCheckService;
	private final DesignQualityCheckResultsReporter designQualityCheckResultsReporter;
	
	public Application(DesignQualityCheckService designQualityCheckService, DesignQualityCheckResultsReporter designQualityCheckResultsReporter) {
		this.designQualityCheckService = designQualityCheckService;
		this.designQualityCheckResultsReporter = designQualityCheckResultsReporter;
	}

	public DesignQualityCheckResults doIt(DesignQualityCheckConfiguration designQualityCheckConfiguration, Printer printer) {
        DesignQualityCheckResults checkResults = designQualityCheckService.analyze(designQualityCheckConfiguration);
        designQualityCheckResultsReporter.report(checkResults, printer);
        
        if (checkResults.failExpectations()) {
        	throw new ThresholdTrespassedException();
        }
        return checkResults;
	}

}

File: src/main/java/org/tindalos/principle/app/service/Application.scala
package org.tindalos.principle.app.service

import org.tindalos.principle.domain.core.AnalysisInput
import org.tindalos.principle.domain.coredetector.AnalysisResult
import org.tindalos.principle.domain.resultprocessing.reporter.Printer

/*
This is the app entry point. Side effects can happen only here in this layer, underneath the code must be pure.
 */
object Application {

  def buildApplication(validateInput: AnalysisInput => (Boolean, String),
                       runAnalysis: AnalysisInput => List[AnalysisResult],
                       makeReports: List[AnalysisResult] => List[(String, Boolean)],
                       printer: Printer) =

    (designQualityCheckConfiguration: AnalysisInput) => {

      val (success, msg) = validateInput(designQualityCheckConfiguration)

      if (success) {

        val analysisResults = runAnalysis(designQualityCheckConfiguration)

        def printReport(report: (String, Boolean)) =
          if (report._2)
            printer.printWarning(report._1)
          else
            printer.printInfo(report._1)

        makeReports(analysisResults) foreach printReport

        (!analysisResults.exists(_.expectationsFailed()), "Expectations failed")

      } else (success, msg)

    }

}